modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -2063,6 +2069,59 @@ void GDScriptAnalyzer::reduce_array(GDScriptParser::ArrayNode *p_array) {
 	p_array->set_datatype(arr_type);
 }
 
+void GDScriptAnalyzer::update_const_expression_builtin_type(GDScriptParser::ExpressionNode *p_expression, const GDScriptParser::DataType &p_type, const char *p_usage, bool p_is_cast) {
+	if (p_expression->get_datatype() == p_type) {
+		return;
+	}
+	if (p_type.kind != GDScriptParser::DataType::BUILTIN && p_type.kind != GDScriptParser::DataType::ENUM) {
+		return;
+	}
+
+	GDScriptParser::DataType expression_type = p_expression->get_datatype();
+	if (!(p_is_cast && p_type.kind == GDScriptParser::DataType::ENUM && p_type.is_meta_type == false && expression_type.builtin_type == Variant::INT) && !is_type_compatible(p_type, expression_type, true, p_expression)) {
+		push_error(vformat(R"(Cannot %s a value of type "%s" as "%s".)", p_usage, expression_type.to_string(), p_type.to_string()), p_expression);
+		return;
+	}
+
+	GDScriptParser::DataType value_type = type_from_variant(p_expression->reduced_value, p_expression);
+	if (expression_type.is_variant() && !is_type_compatible(p_type, value_type, true)) {
+		push_error(vformat(R"(Cannot %s a value of type "%s" as "%s".)", p_usage, value_type.to_string(), p_type.to_string()), p_expression);
+		return;
+	}
+
+	if (p_type.kind == GDScriptParser::DataType::ENUM && value_type.builtin_type == Variant::INT && enum_get_value_name(p_type, p_expression->reduced_value) == StringName()) {
+		if (expression_type.kind == GDScriptParser::DataType::ENUM) {
+			push_error(vformat(R"(Cannot %s "%s.%s" as Enum "%s": no corresponding enum member has the same value (%d).)", p_usage, expression_type.enum_type, enum_get_value_name(expression_type, p_expression->reduced_value), p_type.to_string(), (int64_t)p_expression->reduced_value), p_expression);
```

Comment:
I think this should be a warning, not an error (so as not to create false expectations for the user), since in other cases it is not guaranteed that the value of a variable of type `MyEnum` will be from `MyEnum.values()`.

Reply:
This case happens in casts when a value of one enum casted to represent another.
This is preexisting logic.
Don't understand why would you want it to be a warning instead of an error.
```gdscript
enum A { X = 0 }
enum B { Y = 1 }
var b := A.X as B
```

Reply:
There is no guarantee that a variable of type `MyEnum` cannot contain an arbitrary `int`. It is wrong that in some cases this is an error, and in others a warning.

```gdscript
enum MyEnum {
    VALUE_A,
}

func _run() -> void:
    var x: int = 42
    var a: MyEnum = x # Warning, not an error.
    print(a) # 42
```

Reply:
In your case it will not apply `var x` is not a constant.

Reply:
![](https://user-images.githubusercontent.com/47700418/213986558-70d1f695-eeac-468f-8220-64a4f6218484.png)

Reply:
`42` is a constant in that case. And there is no value 42 in `MyEnum`. How is it not an error?

Reply:
I wrote above. Enums are mostly an alias for `int`. An error **in this case** may give the user the impression that they cannot assign the wrong number **at all**. But this is not true in general.

Reply:
Enum is a quasi-opaque type: an enum value is an int, but an int is not an enum. That's why there is a warning (when we don't know if there is a mismatch) and an error (when we do know that). Enum is not just an alias for int, it has its own type. 

Why type it `MyEnum` and not `int` if you want to assign random values to it?

Reply:
If the variable `a` **can be** equal to `42` (and this is true), then we should not forbid it in some particular cases, while in other cases there will be no error. We need to warn the user that they are doing something wrong (when we can detect it).

Reply:
I mean `a` should not be ever equal to `42` in your example. I guess, there are missing checks in runtime to give an error when `42` is assigned to a variable of type `MyEnum`. I think it is just because enum information is stripped in runtime, maybe it was changed and it is possible to give an error now.

I recommend asking about what enum type represent in developer chat.

Reply:
See also [the comment](https://github.com/godotengine/godot/issues/71763#issuecomment-1399234932).

Reply:
Ok, I asked in developer chat to clarify what does enum type represent.

But if you think that any int can be assigned to an enum type, then why do you think that there should be a warning then?

Reply:
Because it can give users false assumptions about what an enum is in GDScript. Errors should be only for those things that are not allowed in principle. If something is possible but wrong, then it should be a warning, not an error.

Reply:
I mean if it is wrong then you make it an error to make it impossible. Why allow wrong code?

Reply:
I think we don't understand each other. I wrote above several times why I think so.

This is wrong code in the "shouldn't do this" sense, but not in the "can't do this" sense.

Reply:
Done. It is a warning now, not an error.